File: leetcode/101-200/164-maximum-gap/bench/maximum_gap.py
"""LeetCode 164 — Maximum Gap benchmark kernel (Python mirror).

Build-once + punch: LCG-filled N values, call maximum_gap K times perturbing one
element each round. Sink = sum of the K gaps. Identical algorithm to the Kāra /
C / Rust / Go mirrors.
"""


def maximum_gap(nums):
    n = len(nums)
    if n < 2:
        return 0
    lo, hi = nums[0], nums[0]
    for i in range(1, n):
        if nums[i] < lo:
            lo = nums[i]
        if nums[i] > hi:
            hi = nums[i]
    if lo == hi:
        return 0
    bsize = max(1, (hi - lo) // (n - 1))
    bcount = (hi - lo) // bsize + 1
    used = [False] * bcount
    bmin = [0] * bcount
    bmax = [0] * bcount
    for x in nums:
        idx = (x - lo) // bsize
        if used[idx]:
            if x < bmin[idx]:
                bmin[idx] = x
            if x > bmax[idx]:
                bmax[idx] = x
        else:
            bmin[idx] = bmax[idx] = x
            used[idx] = True
    gap = 0
    prev_max = lo
    for i in range(bcount):
        if used[i]:
            if bmin[i] - prev_max > gap:
                gap = bmin[i] - prev_max
            prev_max = bmax[i]
    return gap


def main():
    n = 1000000
    k = 30
    rng = 1000000000
    nums = [0] * n
    state = 12345
    for i in range(n):
        state = (state * 1103515245 + 12345) % 2147483648
        nums[i] = state % rng
    sink = 0
    for round in range(k):
        idx = (round * 7919) % n
        nums[idx] = (nums[idx] + 1 + round) % rng
        sink += maximum_gap(nums)
    print(sink)


main()

File: leetcode/101-200/164-maximum-gap/bench/maximum_gap.rs
// LeetCode 164 — Maximum Gap benchmark kernel (Rust mirror, rustc -O).
//
// Build-once + punch: LCG-filled N values, maximum_gap called K times with a
// one-element perturbation each round. Sink = sum of the K gaps. Identical
// algorithm to the Kāra / C / Go / Python mirrors.

fn maximum_gap(nums: &[i64]) -> i64 {
    let n = nums.len() as i64;
    if n < 2 {
        return 0;
    }
    let mut lo = nums[0];
    let mut hi = nums[0];
    for &x in &nums[1..] {
        if x < lo {
            lo = x;
        }
        if x > hi {
            hi = x;
        }
    }
    if lo == hi {
        return 0;
    }

    let bsize = ((hi - lo) / (n - 1)).max(1);
    let bcount = ((hi - lo) / bsize + 1) as usize;

    let mut used = vec![false; bcount];
    let mut bmin = vec![0i64; bcount];
    let mut bmax = vec![0i64; bcount];

    for &x in nums {
        let idx = ((x - lo) / bsize) as usize;
        if used[idx] {
            if x < bmin[idx] {
                bmin[idx] = x;
            }
            if x > bmax[idx] {
                bmax[idx] = x;
            }
        } else {
            bmin[idx] = x;
            bmax[idx] = x;
            used[idx] = true;
        }
    }

    let mut gap = 0i64;
    let mut prev_max = lo;
    for b in 0..bcount {
        if used[b] {
            if bmin[b] - prev_max > gap {
                gap = bmin[b] - prev_max;
            }
            prev_max = bmax[b];
        }
    }
    gap
}

fn main() {
    let n: i64 = 1_000_000;
    let k: i64 = 30;
    let range: i64 = 1_000_000_000;

    let mut nums = vec![0i64; n as usize];
    let mut state: i64 = 12345;
    for slot in nums.iter_mut() {
        state = (state * 1103515245 + 12345) % 2147483648;
        *slot = state % range;
    }

    let mut sink: i64 = 0;
    for round in 0..k {
        let idx = ((round * 7919) % n) as usize;
        nums[idx] = (nums[idx] + 1 + round) % range;
        sink += maximum_gap(&nums);
    }
    println!("{}", sink);
}
